## Importing necessary libraries

In [151]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.metrics import pairwise_distances
from scipy.spatial.distance import cosine

In [152]:
ls

 Volume in drive C is OS
 Volume Serial Number is E4C2-09B2

 Directory of C:\Users\AKASH KARMAKAR\data science\Excelr assignment code files\13 recommendation system

27-03-2022  17:38    <DIR>          .
25-03-2022  23:07    <DIR>          ..
24-03-2022  00:45    <DIR>          .ipynb_checkpoints
22-03-2022  23:51           536,483 book (1).csv
24-03-2022  00:45           107,540 book_Recommendation_Assignement .ipynb
22-03-2022  23:51                85 Problem Statement (RS).txt
27-03-2022  17:38            82,582 Recommendation system Books Solution.ipynb
24-03-2022  02:34        19,083,230 user.csv
               5 File(s)     19,809,920 bytes
               3 Dir(s)  57,758,986,240 bytes free


## Importing the data

In [153]:
data=pd.read_csv('book (1).csv', encoding='latin')
data.head()

Unnamed: 0  User.ID                                         Book.Title  \
0           1   276726                                Classical Mythology   
1           2   276729                                       Clara Callan   
2           3   276729                               Decision in Normandy   
3           4   276736  Flu: The Story of the Great Influenza Pandemic...   
4           5   276737                             The Mummies of Urumchi   

   Book.Rating  
0            5  
1            3  
2            6  
3            8  
4            6

In [175]:
data[['user_id', 'book_title']].drop_duplicates()

user_id                                         book_title
0      276726                                Classical Mythology
1      276729                                       Clara Callan
2      276729                               Decision in Normandy
3      276736  Flu: The Story of the Great Influenza Pandemic...
4      276737                             The Mummies of Urumchi
...       ...                                                ...
9995   162121       American Fried: Adventures of a Happy Eater.
9996   162121                              Cannibal In Manhattan
9997   162121                    How to Flirt: A Practical Guide
9998   162121                                           Twilight
9999   162129                       Kids Say the Darndest Things

[9993 rows x 2 columns]

## Understanding the data

In [154]:
data.shape

(10000, 4)

In [155]:
data.describe(include='all')

Unnamed: 0        User.ID      Book.Title  Book.Rating
count   10000.00000   10000.000000           10000  10000.00000
unique          NaN            NaN            9659          NaN
top             NaN            NaN  Fahrenheit 451          NaN
freq            NaN            NaN               5          NaN
mean     5000.50000   95321.249800             NaN      7.56630
std      2886.89568  117645.703609             NaN      1.82152
min         1.00000       8.000000             NaN      1.00000
25%      2500.75000    2103.000000             NaN      7.00000
50%      5000.50000    3757.000000             NaN      8.00000
75%      7500.25000  162052.000000             NaN      9.00000
max     10000.00000  278854.000000             NaN     10.00000

In [156]:
data['Book.Title'].nunique()

9659

## Summary
- There are 9659 books out of 10000 books are unique
- There are 2182 users. 
- Fahreinheit 451 book has the highest frequency of 5

In [157]:
plt.style.use('seaborn-darkgrid')
plt.figure(figsize=(10,6))
data["book_rating"].value_counts().sort_values(ascending=True).plot(kind="bar")
plt.xticks(rotation=0)
plt.xlabel('Ratings')
plt.ylabel('Freq')
plt.show()

KeyError: 'book_rating'

<Figure size 720x432 with 0 Axes>

In [158]:
data.rename({'Unnamed: 0': 'user_index', 'User.ID':'user_id', 'Book.Title': 'book_title', 'Book.Rating': 'book_rating'},axis=1, inplace=True)

In [159]:
data.head()

user_index  user_id                                         book_title  \
0           1   276726                                Classical Mythology   
1           2   276729                                       Clara Callan   
2           3   276729                               Decision in Normandy   
3           4   276736  Flu: The Story of the Great Influenza Pandemic...   
4           5   276737                             The Mummies of Urumchi   

   book_rating  
0            5  
1            3  
2            6  
3            8  
4            6

In [160]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   user_index   10000 non-null  int64 
 1   user_id      10000 non-null  int64 
 2   book_title   10000 non-null  object
 3   book_rating  10000 non-null  int64 
dtypes: int64(3), object(1)
memory usage: 312.6+ KB


In [161]:
 data.isna().sum()

user_index     0
user_id        0
book_title     0
book_rating    0
dtype: int64

- There is no null value in the data

## Data mining

In [162]:
user_book_df = data.pivot_table(index = 'user_id',
                                 columns = 'book_title',
                                 values = 'book_rating').reset_index(drop=True)

user_book_df


book_title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
0                            NaN                                         NaN   
1                            NaN                                         NaN   
2                            NaN                                         NaN   
3                            NaN                                         NaN   
4                            NaN                                         NaN   
...                          ...                                         ...   
2177                         NaN                                         NaN   
2178                         NaN                                         NaN   
2179                         NaN                                         NaN   
2180                         NaN                                         NaN   
2181                         NaN                                         NaN   

book_title   Repairing PC Drives &amp  '48  \
0                                 NaN  NaN   
1                                 NaN  NaN   
2                                 NaN  NaN   
3                                 NaN  NaN   
4                                 NaN  NaN   
...                               ...  ...   
2177                              NaN  NaN   
2178                              NaN  NaN   
2179                              NaN  NaN   
2180                              NaN  NaN   
2181                              NaN  NaN   

book_title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
0                                                         NaN                       
1                                                         NaN                       
2                                                         NaN                       
3                                                         NaN                       
4                                                         NaN                       
...                                                       ...                       
2177                                                      NaN                       
2178                                                      NaN                       
2179                                                      NaN                       
2180                                                      NaN                       
2181                                                      NaN                       

book_title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
0                                                         NaN              
1                                                         NaN              
2                                                         NaN              
3                                                         NaN              
4                                                         NaN              
...                                                       ...              
2177                                                      NaN              
2178                                                      NaN              
2179                                                      NaN              
2180                                                      NaN              
2181                                                      NaN              

book_title  01-01-00: A Novel of the Millennium  \
0                                           NaN   
1                                           NaN   
2                                           NaN   
3                                           NaN   
4                                           NaN   
...                                         ...   
2177                                        NaN   
2178                                        NaN   
2179                                        NaN   
2180                                      

In [163]:
user_book_df.fillna(0, inplace=True)

In [164]:
user_book_df

book_title   Jason, Madison &amp   Other Stories;Merril;1985;McClelland &amp  \
0                            0.0                                         0.0   
1                            0.0                                         0.0   
2                            0.0                                         0.0   
3                            0.0                                         0.0   
4                            0.0                                         0.0   
...                          ...                                         ...   
2177                         0.0                                         0.0   
2178                         0.0                                         0.0   
2179                         0.0                                         0.0   
2180                         0.0                                         0.0   
2181                         0.0                                         0.0   

book_title   Repairing PC Drives &amp  '48  \
0                                 0.0  0.0   
1                                 0.0  0.0   
2                                 0.0  0.0   
3                                 0.0  0.0   
4                                 0.0  0.0   
...                               ...  ...   
2177                              0.0  0.0   
2178                              0.0  0.0   
2179                              0.0  0.0   
2180                              0.0  0.0   
2181                              0.0  0.0   

book_title  'O Au No Keia: Voices from Hawai'I's Mahu and Transgender Communities  \
0                                                         0.0                       
1                                                         0.0                       
2                                                         0.0                       
3                                                         0.0                       
4                                                         0.0                       
...                                                       ...                       
2177                                                      0.0                       
2178                                                      0.0                       
2179                                                      0.0                       
2180                                                      0.0                       
2181                                                      0.0                       

book_title  ...AND THE HORSE HE RODE IN ON : THE PEOPLE V. KENNETH STARR  \
0                                                         0.0              
1                                                         0.0              
2                                                         0.0              
3                                                         0.0              
4                                                         0.0              
...                                                       ...              
2177                                                      0.0              
2178                                                      0.0              
2179                                                      0.0              
2180                                                      0.0              
2181                                                      0.0              

book_title  01-01-00: A Novel of the Millennium  \
0                                           0.0   
1                                           0.0   
2                                           0.0   
3                                           0.0   
4                                           0.0   
...                                         ...   
2177                                        0.0   
2178                                        0.0   
2179                                        0.0   
2180                                      

## checking the user similarity using cosine metric

In [165]:

us_sim = 1-pairwise_distances(user_book_df, metric='cosine' )

#storing the values in a dataframe
us_sim_df=pd.DataFrame(us_sim)

#converting the user indexes into the user id
us_sim_df.index=data.user_id.unique()
us_sim_df.columns=data.user_id.unique()

us_sim_df.head()

276726  276729  276736  276737  276744  276745  276747  276748  \
276726     1.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     1.0     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0     1.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0     0.0     1.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0     0.0     0.0     1.0     0.0     0.0     0.0   

        276751  276754  ...  162085  162091  162092  162095  162103  162107  \
276726     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

        162109  162113  162121  162129  
276726     0.0     0.0     0.0     0.0  
276729     0.0     0.0     0.0     0.0  
276736     0.0     0.0     0.0     0.0  
276737     0.0     0.0     0.0     0.0  
276744     0.0     0.0     0.0     0.0  

[5 rows x 2182 columns]

In [166]:
us_sim_df.shape

(2182, 2182)

- We can see that the diagonal values are 1, it may cause error while performing the recommendation , we will remove them.

In [167]:
np.fill_diagonal(us_sim, 0)
us_sim_df.head()

276726  276729  276736  276737  276744  276745  276747  276748  \
276726     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0     0.0     0.0     0.0     0.0     0.0     0.0   

        276751  276754  ...  162085  162091  162092  162095  162103  162107  \
276726     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276729     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276736     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276737     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   
276744     0.0     0.0  ...     0.0     0.0     0.0     0.0     0.0     0.0   

        162109  162113  162121  162129  
276726     0.0     0.0     0.0     0.0  
276729     0.0     0.0     0.0     0.0  
276736     0.0     0.0     0.0     0.0  
276737     0.0     0.0     0.0     0.0  
276744     0.0     0.0     0.0     0.0  

[5 rows x 2182 columns]

- Now we have all pure simillarity values

In [168]:
# now we can see which users are correlated.
us_sim_df.idxmax(axis=1)


276726    276726
276729    276726
276736    276726
276737    276726
276744    276726
           ...  
162107    276726
162109    276726
162113    161453
162121    276726
162129    276726
Length: 2182, dtype: int64

- Extract the books which user id 277128 & 277196 have read

In [169]:
us_sim_df.loc[162113, 161453]

0.023463636182787573

In [170]:
data[data['user_id']==276729]

user_index  user_id            book_title  book_rating
1           2   276729          Clara Callan            3
2           3   276729  Decision in Normandy            6

In [171]:
data[data['book_title']=='Clara Callan'] 

user_index  user_id    book_title  book_rating
1           2   276729  Clara Callan            3

In [172]:
a=data[(data['user_id']==162113) | (data['user_id']==161453)]
a

user_index  user_id                                         book_title  \
8959        8960   161453  Bread, Tomato, Garlic: Quick Cooking With 3 Ma...   
8960        8961   161453  The Ubiquitous Shrimp: From Simple to Exotic, ...   
9989        9990   162113       The Cape Ann (Contemporary American Fiction)   

      book_rating  
8959            9  
8960            8  
9989            8

In [114]:
user1=data[(data['user_id']==277128)]
user2=data[(data['user_id']==277196)]

In [115]:
type(user2)

pandas.core.frame.DataFrame

In [117]:
user1['book_title']

436                       L'home del sac
437    Gràcies per la propina (Columna)
Name: book_title, dtype: object

In [118]:
user2['book_title']

514           From the Earth: Chinese Vegetarian Cooking
515    Living in the Presence: Disciplines for the Sp...
Name: book_title, dtype: object

In [123]:
pd.merge(user1, user2, on= 'book_title', how='outer')

user_index_x  user_id_x                                         book_title  \
0         437.0   277128.0                                     L'home del sac   
1         438.0   277128.0                  Gràcies per la propina (Columna)   
2           NaN        NaN         From the Earth: Chinese Vegetarian Cooking   
3           NaN        NaN  Living in the Presence: Disciplines for the Sp...   

   book_rating_x  user_index_y  user_id_y  book_rating_y  
0            6.0           NaN        NaN            NaN  
1            8.0           NaN        NaN            NaN  
2            NaN         515.0   277196.0           10.0  
3            NaN         516.0   277196.0            7.0

## user 1 books will be recommended to user 2 & user 2 books will be recommended to user 1